### ISSUE TITLE
The ElasticsearchSource using Akka HTTP Does Not Support TLS (or the setting of headers)

### ISSUE THREAD

**Issue by jabbaugh** [akka/alpakka#2645]
### Short description

The move for the ElasticsearchSource to use Akka HTTP (instead of passing in a RestClient) does not support TLS authentication or headers to be inserted into the HttpRequest.

### Details

The change that removed the implicit RestClient from the ElasticsearchSource to now use Akka Http no longer supports TLS auth. While you can pass in credentials using the ElasticsearchConnectionSettings.withCredentials that is only used in BasicAuth (see ElasticsearchAPI).

An approach that would include the ability to add headers (ElasticsearchConnectionSettings.withHeaders) would allow more flexibility for non BasicAuth or standard requests.

The header then could be added to the HttpRequest in the ElasticsearchSourceLogic.sendScrollScanRequest method.

#### Another Option
Alternatively, you could add the ability to pass in a HttpsConnectionContext (e.g. ElasticsearchConnectionSettings.withConnectionContext) and then use that context on the singleRequest in ElasticsearchApi.executeRequest method (e.g. http.singleRequest(request, connectionSettings.connectionContext.get)

**Comment by goldmar** [comment]
I have run into the same issue. In my opinion, the best approach would be to pass in `HttpExt` since some configuration can only be done on that level. For instance, we need the ability to [disable hostname verification](https://doc.akka.io/docs/akka-http/current/client-side/client-https-support.html).

Currently, `HttpExt` is being instantiated in `ElasticsearchFlow.basicStageFlow`; maybe it can be passed in as an implicit parameter in the public API?

**Comment by michaeljmarshall** [comment]
@seglo - just want to raise this as an issue I'm hoping is fixed before the 3.0.0 release. Whatever the implementation is, it'll likely introduce a (possibly breaking) change to the public API.

> the best approach would be to pass in HttpExt

I think passing in the `HttpExt` solution would require setting a default https context for the whole http extension, which I am pretty sure is static for the whole actor system.

In looking at the `singleRequest` method signature:

```scala
  def singleRequest(
    request:           HttpRequest,
    connectionContext: HttpsConnectionContext = defaultClientHttpsContext,
    settings:          ConnectionPoolSettings = defaultConnectionPoolSettings,
    log:               LoggingAdapter         = system.log): Future[HttpResponse]
```

I noticed that it allows for passing in the `HttpsConnectionContext`. In that case, the `executeRequest` method could be updated to use a parameterized `HttpsConnecgtionContext` (this method is new to this release so that alpakka elasticsearch uses akka http instead of the elasticsearch rest client).

Here is that method currently:

```scala
  def executeRequest(
      request: HttpRequest,
      connectionSettings: ElasticsearchConnectionSettings
  )(implicit http: HttpExt): Future[HttpResponse] = {
    if (connectionSettings.hasCredentialsDefined) {
      http.singleRequest(
        request.addCredentials(BasicHttpCredentials(connectionSettings.username.get, connectionSettings.password.get))
      )
    } else {
      http.singleRequest(request)
    }
  }
```

Perhaps it might make sense to update the `ElasticsearchConnectionSettings` case class to include more security configurations?

**Comment by jabbaugh** [comment]
I created a PR that adds both Headers as well as the HttpsConnectionContext (see https://github.com/akka/alpakka/pull/2652)

A different approach would be to add TLS creds to the ElasticsearchConnectionSettings and support both BasicAuth as well as TLS creds. It is possible to set the TLS creds on the HttpsConnectionContext.

**Comment by goldmar** [comment]
@michaeljmarshall you are right, the `HttpExt` is used globally across the actor system. I was not aware of that. That means changing the default settings is already possible without any changes. Of course it would be better to be able to configure only the requests to Elasticsearch. The PR by @jabbaugh looks good to me.

**Comment by jabbaugh** [comment]
The PR (https://github.com/akka/alpakka/pull/2652) passes all validation and now only needs approval by a reviewer with write access.

**Comment by seglo** [comment]
I believe this has been addressed with #2652. Correct me if I'm wrong.
